v2: list dependencies from either import path or go binary

Bobgy opened the pull request:
This is the first PR of https://github.com/google/go-licenses/issues/70

Bobgy commented:
Hi @wlynch! Here's the first PR

Bobgy commented:
Todo:
* Rm v2/deps/main
* Include third-party lichen module

Bobgy commented:
All the clean ups are done. Ready for code review : )

cc @wlynch @06kellyjac

Bobgy commented:
I'm not cleaning up the commits, because I always use squash merge when merging a PR.
Please let me know if this project has a different convention.

Bobgy commented on v2/deps/go_binary.go:
```diff
@@ -0,0 +1,101 @@
+	ImportPath string
+	// version, example: v1.2.3, v0.0.0-20201021035429-f5854403a974
+	Version string
+}
```
Done

wlynch commented on v2/gocli/go_binary.go:
```diff
@@ -0,0 +1,111 @@
+			err = fmt.Errorf("listModulesInGoBinary(path=%q): %w", path, err)
+		}
+	}()
+	depsBuildInfo, err := lichenmodule.Extract(context.Background(), path)
```
If we aren't sure what the context is being used for, it might be safer to just taken in a context arg and pass it through.

wlynch commented on v2/gocli/list.go:
```diff
@@ -0,0 +1,86 @@
+}
+
+// Helper to build a module path -> module dict.
+func BuildModuleDict(modules []Module) map[string]Module {
```
Looks like this is only used in go_binary.go. Let's move this there and unexport.

Bobgy replied:
Let me double check if I got other usages in unsent PRs.

Bobgy replied:
You are right, in fact I always use the module dict, so I merged two methods together.

wlynch commented on v2/README.md:
```diff
@@ -0,0 +1,223 @@
+# go-licenses
```
Looks like this needs to be updated to reflect the current state.

Bobgy replied:
I have some more changes for readme in my fork's master branch.

I intend to do a batch update later because it will be some throw away work only to sync the readme with state of the PR, if we do it now.

What do you think?

wlynch replied:
Let's try to keep this as up to the date to the current code as possible. Feel free to keep the current version in a fork and import back in later! :D

Bobgy replied:
That makes sense!
I updated README to only include information related to current state.

Bobgy commented on v2/gocli/go_binary.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func joinModulesMetadata(refs []model.ModuleReference) (modules []Module, err error) {
+	localModules, err := ListModules()
```
Yes, I am taking that as a prerequisite. It works for my use case, because I built the binary, so I always have access to the module dir. I am curious when you wouldn't.

I see your point for making it also accept package path, so it could be backwards compatible. I think that's a good idea. I will put it on the roadmap as a P1. I prefer getting the end to end flow working before introducing options. What do you think?

I didn't quite understand one point:
> since we could go from binary path -> module path -> packages

How do we get packages from module path if we are not already in the module dir? Do you mean scenarios like the module itself lives in GOPATH?

Bobgy replied:
Added in roadmap: https://github.com/google/go-licenses/issues/70#issuecomment-864388550

wlynch replied:
2 use cases come to mind -

1. CI jobs that pull in multiple repos - the modules would be in subdirectories and might not start from a module directory.
2. If you want to query for just the licenses of the binary (without the module).

I think it's clearer to have users specify `go-licenses check path/to/binary ./...`. I'm worried that baking in the assumption that users will always execute the tool from a module directory may not work for all use cases.

wlynch replied:
Ah - I just meant that for a given arg we can try to parse it as a binary first, get the modules, then use the same packages.Load that we're already using to look up the deps.

Bobgy replied:
I think there are some complexities in getting the same information outside of the go module, which makes me prefer not supporting this feature initially. Let me confirm my understanding, when outside of a go module, we can only resolve packages in GOPATH mode, so is "packages.Load". Is that right? Therefore, whenever possible I think people should avoid doing that.

Complexities:
1. versioning, outside of the go module, we cannot go get a module with version. (major concern)
2. [replace directives](https://golang.org/ref/mod#go-mod-file-replace), a replaced module may be a local path too, in which case, you can only know what code exactly was when you are checking licenses in the go module. (probably not a strong argument, because this is rather corner case.)

> 2 use cases come to mind -
> 
> 1. CI jobs that pull in multiple repos - the modules would be in subdirectories and might not start from a module directory.

Because of complexities mentioned above, I think for the same reasons, to run commands like `go build`, `go test`, you would also need to cd into the module directory. I feel like it's fair to assume we need the same prerequisite when checking licenses.

But I think you reminded me that, we do not need to require current working dir is the root of the module. We should be able to run it in sub folders just like other go CLI commands.

> 2. If you want to query for just the licenses of the binary (without the module).
> 
> I think it's clearer to have users specify `go-licenses check path/to/binary ./...`. I'm worried that baking in the assumption that users will always execute the tool from a module directory may not work for all use cases.

I think this would be a useful scenario, just that I cannot understand how we can download source code of all dependencies without the module directory?
I can think of probably two ways:
1. resolve from GOPATH -- this also assumes sth on user environment and there can be version skew. So I'd refrain from doing this
2. construct a temporary go module with a go.mod including all the modules we found in the binary, and then use the dummy go module env to download all source code and check dependencies. This probably works, but seems hacky too.

Bobgy replied:
Let me confirm, do you think this is a blocker for this PR?

wlynch replied:
Did some testing of the existing behavior - looks like this is influenced by the `GO111MODULE` environment variable. 

You're right that even with the GOPATH mode, we still need to run within the GOPATH for things to work, and that this is consistent with other go tools, and even the existing go-licenses tool (with `GO111MODULE=on` go-licenses needs to run within a module directory to function properly). So LGTM for requiring this.

I still want to push back on automatically including everything in the current module directory by default. In the same way you need to distinguish check 1 package (`go-licenses check .`) vs check all packages (`go-licenses ./...`), I'd like to do something similar for binaries as well. Something else I realized in messing with things is that `go version` already recursively traverses directories for binaries, so I'm wondering how we can make a better distinction between the package `.` vs "look for binaries in this directory" `.`. 2 possible ideas -

1. Put binary commands into their own command group similar to `go list` vs `go version`
  
  ```sh
  $ go-licenses binary save .
  ```

2. Put binary search path into a flag

  ```sh
  $ go-licenses binary save --binary_path=. ./...
  ```

Let me know what you think!

Bobgy replied:
> So LGTM for requiring this.

Great, we agreed on one more thing.

> I still want to push back on automatically including everything in the current module directory by default.

I guess my function naming can be a little misleading? `join` --> `inner join`.
I am already doing what you mentioned -- reading all modules in current workdir, but I do not add all of them by default. Only modules used in the go binary are returned.

> Something else I realized in messing with things is that go version already recursively traverses directories for binaries, so I'm wondering how we can make a better distinction between the package . vs "look for binaries in this directory" .

Interesting, so far my use-case is limited to getting licenses for one binary. Let me think about what's a better way to handle the ambiguity.

Anyway, I think this PR only implements listing modules in a binary. We can keep discussing what's the best CLI commands + semantics in upcoming PRs.

Bobgy commented on v2/gocli/go_binary_test.go:
```diff
@@ -0,0 +1,95 @@
+		t.Fatal(err)
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("List modules in go binary built from %s", tc.workdir), func(t *testing.T) {
```
Thanks! Good to learn this

Bobgy commented on v2/gocli/list.go:
```diff
@@ -0,0 +1,86 @@
+	"os/exec"
+)
+
+// Go module metadata
+type Module struct {
+	Path      string // go import path
+	Main      bool   // whether this is the main module in the workdir
+	Version   string
+	Time      string
+	Indirect  bool
+	Dir       string
+	GoMod     string
+	GoVersion string
+	Replace   *Module
+}
```
Nice, didn't know where it was!
Will change

Bobgy commented on v2/tests/modules/cli02/cmd/root.go:
```diff
@@ -0,0 +1,85 @@
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	homedir "github.com/mitchellh/go-homedir"
+	"github.com/spf13/viper"
```
Note, this folder is generated by cobra CLI without change.
I only use it for testing getting go modules

wlynch replied:
I like the idea of making this very simple module/binary that obviously does nothing. It will help make it easier for new readers to understand what this does - even though this is the default generated cobra code, you have to read through it to figure that out.

Having an empty main with an `_` import is functionally equivalent, but much simpler to grok if you have no background on this file.

Bobgy commented on v2/third_party/uw-labs/lichen/METADATA:
```diff
@@ -0,0 +1,16 @@
+description:
+    "uw-labs/lichen is a license tool to analyze go binaries"
+
+third_party {
```
Yes, I am using the internal package, but I didn't make any changes.

The file itself follows google third party metadata syntax and it's manually maintained. We don't necessarily need to be so strict with the format. I think it does cause some confusion.

I will just change its format to markdown to simplify.

wlynch replied:
Re forking the library - I'm hesitant to fork since it makes the dependency upgrade process much more manual. Looking at the lichen code in more detail, IIUC it looks like this is effectively wrapping `go version -m`. If that's all we need, it might be simpler to just have our own func that accomplishes a similar result (I think we could use the code you already wrote here actually).

Re: third_party format - we should avoid using this the schema isn't widely accessible (it would be difficult for non-googlers to know what is expected here).

Bobgy commented on v2/gocli/list.go:
```diff
@@ -0,0 +1,86 @@
+		//         "GoMod": "/home/gongyuan_kubeflow_org/go/pkg/mod/cache/download/k8s.io/kubernetes/@v/v1.11.1.mod"
+		// }
+		// handle replace directives
+		if m.Replace != nil {
+			m = *m.Replace
+		}
```
The version is wrong 😔. I recall that there were some other field that made me do this. Cannot recall right now.

Bobgy replied:
Added a comment to clarify, hope that helps

Bobgy commented:
Hi @wlynch, thank you for the detailed review!
I can see a big quality improvement with your comments.
I've addressed all comments except https://github.com/google/go-licenses/pull/71#discussion_r658264610 (which I want to defer to future work to keep the scope of this PR controllable.)

It's now ready for another round of reviews when you have time.

wlynch reviewed (commented):
Some of these comments refernce changes that aren't reflected in the PR. Is there a commit that needs to be pushed?
:)

Thanks!

Bobgy commented on v2/gocli/list.go:
```diff
@@ -0,0 +1,86 @@
+
+// List go modules with metadata using go CLI list command.
+func ListModules() ([]Module, error) {
+	out, err := exec.Command("go", "list", "-m", "-json", "all").Output()
```
being discussed in https://github.com/google/go-licenses/pull/71#discussion_r658264610

Bobgy commented on v2/README.md:
````diff
@@ -0,0 +1,223 @@
+```bash
+curl -LO download-url/go-licenses-linux.tar.gz
+tar xvf go-licenses-linux.tar.gz
+sudo mv go-licenses/* /usr/local/bin/
````
Removed this section in the README.
I'll try to build `go get` support after code in the repo can work end to end.

Bobgy commented:
Strange, I never force pushed anything into the branch. Never saw uncommentable comments before.

Bobgy commented:
> I like the idea of making this very simple module/binary that obviously does nothing. It will help make it easier for new readers to understand what this does - even though this is the default generated cobra code, you have to read through it to figure that out.
> 
> Having an empty main with an _ import is functionally equivalent, but much simpler to grok if you have no background on this file.

These are great points! I'll change according to your suggestions.

Bobgy commented:
> Re forking the library - I'm hesitant to fork since it makes the dependency upgrade process much more manual. Looking at the lichen code in more detail, IIUC it looks like this is effectively wrapping go version -m. If that's all we need, it might be simpler to just have our own func that accomplishes a similar result (I think we could use the code you already wrote here actually).

Yes, you understandings are correct. The code I forked didn't do much, it's just a thin wrapper over `go version -m` CLI command. Therefore, I don't think we need to consider upgrade in any near future until go decides to change output format of the CLI argument. so I don't think this is a high priority thing

However, there are serious copyright problems if I just move the code out of third_party, because I only modified some code, most of the code were written by uw-labs people. To avoid legal issues, I'll have to rewrite this piece of code without looking at the original code (which I've read once). I don't think that's a valuable use of our time at the moment, we can maybe do that later.

Were there other strong reasons to avoid this? I can see that using any third_party folder can create some trouble for importing the code here into Google. I'd like to defer this problem until we have an immediate concern.

> Re: third_party format - we should avoid using this the schema isn't widely accessible (it would be difficult for non-googlers to know what is expected here).

Agreed, I've removed the METADATA file in the first iteration. Were you reading an outdated version?

Bobgy commented:
TODOs:
* [ ] use packages.load instead of go CLI
* [ ] make the test code empty using underscore imports

Bobgy commented:
@wlynch I tried using `packages.Load` instead, but I hit issue: https://github.com/google/go-licenses/issues/18

here's my implementation:
```
func loadModulesMetadata(refs []model.ModuleReference) (modules []packages.Module, err error) {
	modules = make([]packages.Module, 0)
	for _, ref := range refs {
		pkgs, err := packages.Load(&packages.Config{
			Mode: packages.NeedModule,
		}, ref.Path)
		if err != nil {
			return nil, err
		}
		if len(pkgs) != 1 {
			return nil, fmt.Errorf("got len(pkgs)=%v, expect 1", len(pkgs))
		}
		pkg := pkgs[0]
		if len(pkg.Errors) > 0 {
			log.Printf("got errs: %+v", pkg.Errors)
			continue
		}
		if pkg.Module == nil {
			log.Printf("got nil module: %+v, path=%s", pkg, pkg.PkgPath)
			continue
		}
		modules = append(modules, *pkg.Module)
	}
	return modules, nil
}
```

And I got the error when running unit test for cli02 test module:
```
2021/07/31 12:25:59 got errs: [-: no required module provides package golang.org/x/sys; to add it:
	go get golang.org/x/sys]
--- FAIL: TestListModulesInGoBinary (2.12s)
    --- FAIL: TestListModulesInGoBinary/../tests/modules/cli02 (2.00s)
        /home/gongyuan_kubeflow_org/github/google/go-licenses/v2/gocli/go_binary_test.go:91: 
            	Error Trace:	go_binary_test.go:91
            	Error:      	Not equal: 
            	            	expected: []string{"github.com/fsnotify/fsnotify", "github.com/hashicorp/hcl", "github.com/magiconair/properties", "github.com/mitchellh/go-homedir", "github.com/mitchellh/mapstructure", "github.com/pelletier/go-toml", "github.com/spf13/afero", "github.com/spf13/cast", "github.com/spf13/cobra", "github.com/spf13/jwalterweatherman", "github.com/spf13/pflag", "github.com/spf13/viper", "github.com/subosito/gotenv", "golang.org/x/sys", "golang.org/x/text", "gopkg.in/ini.v1", "gopkg.in/yaml.v2"}
            	            	actual  : []string{"github.com/fsnotify/fsnotify", "github.com/hashicorp/hcl", "github.com/magiconair/properties", "github.com/mitchellh/go-homedir", "github.com/mitchellh/mapstructure", "github.com/pelletier/go-toml", "github.com/spf13/afero", "github.com/spf13/cast", "github.com/spf13/cobra", "github.com/spf13/jwalterweatherman", "github.com/spf13/pflag", "github.com/spf13/viper", "github.com/subosito/gotenv", "golang.org/x/text", "gopkg.in/ini.v1", "gopkg.in/yaml.v2"}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,2 +1,2 @@
            	            	-([]string) (len=17) {
            	            	+([]string) (len=16) {
            	            	  (string) (len=28) "github.com/fsnotify/fsnotify",
            	            	@@ -14,3 +14,2 @@
            	            	  (string) (len=26) "github.com/subosito/gotenv",
            	            	- (string) (len=16) "golang.org/x/sys",
            	            	  (string) (len=17) "golang.org/x/text",
            	Test:       	TestListModulesInGoBinary/../tests/modules/cli02
```

I think the root cause is that golang.org/x/sys is a module, but it is not a package.
Refer to its source code: https://cs.opensource.google/go/x/sys.

So we cannot load the package golang.org/x/sys.

Bobgy commented:
Therefore, I think I'd prefer staying with existing implementation.

wlynch commented on v2/gocli/go_binary.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+func joinModulesMetadata(refs []model.ModuleReference) (modules []packages.Module, err error) {
+	localModulesDict, err := ListModules()
```
Let's add a comment explaining why/how we're using the local modules data.

Bobgy replied:
Done

wlynch commented on v2/gocli/go_binary.go:
```diff
@@ -0,0 +1,111 @@
+		if localModule.Dir == "" {
+			return nil, fmt.Errorf("Module %v's local directory is empty. Did you run go mod download?", ref.Path)
+		}
+		if localModule.Version != ref.Version {
```
This causes cause problems if the current module context doesn't match the same version as the binary. I think we'll need to rely on `go list -m` (or some library equivalent if available) instead.

To test this I used a modified cli02 binary using `golang.org/x/tools@v0.1.3`, with a go-licenses@v2 using `golang.org/x/tools@v0.1.4` ([diff](https://github.com/Bobgy/go-licenses/compare/pr1...wlynch:pr-71-wlynch))

```sh
$ go test ./...
--- FAIL: TestListModulesInGoBinary (0.28s)
    --- FAIL: TestListModulesInGoBinary/../tests/modules/cli02 (0.13s)
        go_binary_test.go:81: Found golang.org/x/tools v0.1.3 in go binary, but v0.1.4 is downloaded in go modules. Are you running this tool from the working dir to build the binary you are analyzing?
FAIL
$ go list -json -m golang.org/x/tools@v0.1.3  # Even though last command failed, module is downloaded locally
{
 "Path": "golang.org/x/tools",
 "Version": "v0.1.3",
 "Time": "2021-06-09T21:40:20Z",
 "Dir": "/usr/local/google/home/wlynch/pkg/mod/golang.org/x/tools@v0.1.3",
 "GoMod": "/usr/local/google/home/wlynch/pkg/mod/cache/download/golang.org/x/tools/@v/v0.1.3.mod",
 "GoVersion": "1.17"
}
```

Bobgy replied:
Wow, I didn't know we could use `go list -json -m <module@version>` to also download the module and get metadata at the same time. And I just tried, it's not restricted by the current module dir. I can run the command anywhere to get the modules.
I think we are now very close to lift the restriction of go module dir, if we just `go list -json -m <the-entire-list-of-modules>` to get all the modules.

There's one remaining challenge to me, the main module to build a binary is usually not versioned. How can we know what it is if outside of the go working dir?

> This causes cause problems if the current module context doesn't match the same version as the binary.

For clarification, it was expected behavior in my design. How do we know where the main module is and whether it is the correct version? If we are in a working dir and found go modules version has a mismatch, a more serious problem is that most likely the main module's version is incorrect

Bobgy replied:
> The binary module is going to be different from the local working directory module, unless the local working directory module is the module that produces the binary (but if that were the case, my expectation is you would run go-licenses on the source package itself rather than the binary)

My use-case is exactly running go-licenses in the same working directory that produces the binary, that's a prerequisite for running the tool. 
The reason I chose to run the tool on the binary instead of source code is because some go module dependencies (test / tool dependencies) are trimmed when compiling to the binary, so we can check license only for the modules included in the binary.

When you said running go-licenses on the source package itself, I assume you are talking about using packages.Load like the same logic in go-licenses v1. I think both solutions work, if you have strong opinions, I can try to use packages.Load instead.

I personally feel that it's better to rely on the modules metainfo from built binary, because it's better for single source of truth -- the go compiler checks for all dependencies, compile to a binary and record the modules metainfo in the binary.
On the contrary, if we use packages.Load to get all the modules from a package, the go compiler will collect all dependencies once, and packages.Load will get all the modules for the second time. How do we guarantee the two are giving the same results? I think there are potential risks for:
* version skew -- when go compiler has a different version from packages.Load
* build tag / config skew -- go compiler uses different build tags / flags than what packages.Load is using
If we read the modules meta info directly from the binary, these risks are removed, because only the go compiler decides what modules are pulled in.

I believe we are making a trade off here.

Option 1 - the tool only runs in the same workdir used to build the binary.
Option 2 - the tool only works for binaries built by `go install module@version`.

Considering the major user scenario should be module authors adding license info to binaries built by themselves, I think option 1 is better.

After some experimentation using the new tool in [github.com/kubeflow/pipelines](github.com/kubeflow/pipelines), my typical licensing management workflow is like the following:
1. keep a copy of generated license csv file in the repo
2. in presubmit test, besides running go unit tests, also run `go-licenses csv` and verify the license csv is up-to-date. When there's an error, the PR author should check licenses for updated dependencies and update the licenses csv.
3. when releasing, `go-licenses save` the notices to the container image containing the binary

In all the scenarios, the person/CICD tool running go-licenses should be using it in exactly the same workdir as where the binary is built.

wlynch replied:
Thanks for the additional context! I was under the assumption you wanted to include additional binary tools as part of the license output, or to generate a license summary from a standalone unrelated binary. IIUC, you want to use the binary metadata as a shortcut to generate a minimal license summary that only includes dependencies that would be included in binaries that you are distributing.

> The reason I chose to run the tool on the binary instead of source code is because some go module dependencies (test / tool dependencies) are trimmed when compiling to the binary, so we can check license only for the modules included in the binary. 

Test dependencies are already excluded unless `Test: true` is included in [packages.Config](https://pkg.go.dev/golang.org/x/tools@v0.1.5/go/packages#Config) (which is [currently not enabled](https://github.com/google/go-licenses/blob/3099c18c36e1e78150f562c2a2d99e916b43b009/licenses/library.go#L69-L72) for go-licenses). Related: https://github.com/google/go-licenses/issues/62

Tool dependencies (assuming this is following https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module) would be treated like any other dependency and only be present if they are included transitively.

I think this would cover most cases to generate a minimal license summary, but if you have any examples in your project that don't fit well let me know! Would happy to dive deeper here.

> version skew -- when go compiler has a different version from packages.Load

I'm not particularly worried about go compiler vs packages.Load version skew. Go has a pretty good track record when it comes to backwards compatibility. packages.Load is used by other tools in golang.org/x/tools, so I would imagine/hope any breaking changes / incompatibilities would be noticed prior to major Go releases.

What I am more worried about is version skew between the code and the binary, since the binary could have been built from a different version than what is currently present in the repo / go.mod. I'd rather lean on the code as the source of truth as much as possible, since that's what `go.mod` is referring to.

> build tag / config skew -- go compiler uses different build tags / flags than what packages.Load is using

Although go-licenses does not directly accept build flags on its own CLI, we do support build flags via [GOFLAGS](https://pkg.go.dev/cmd/go#hdr-Environment_variables).
If we wanted, we could pass through flags via `packages.Config`.

> Considering the major user scenario should be module authors adding license info to binaries built by themselves, I think option 1 is better.

Agreed - I also prefer Option 1, but would prefer to use the source code as the source of truth rather than the binary.

FWIW - projects I'm involved in follows an identical pattern to what you're describing in your release workflow using the existing tool.

Bobgy replied:
Hello @wlynch, I think the new update will be what you want.
As we discussed initially, we can support both getting dependencies from the binary or from code + import path.

I've included another implementation to get all the dependencies and their modules using packages.Visit.

> FWIW - projects I'm involved in follows an identical pattern to what you're describing in your release workflow using the existing tool.

Glad to know, that gives me more confidence of the UX.

Bobgy replied:
Non-P0 idea:

By the way, after thinking a bit more about the use-case for analyzing completely unknown go binaries.
I think it can be a good additional feature.
As discussed in https://github.com/google/go-licenses/pull/71#discussion_r691427308, the current limitation is that we do not know the exact version of the main module. However, the limitations seem to be a good trade off.

e.g. consider this UX:

I use go-licenses to analyze random binaries taken from containers, go-licenses can get exact version of all the dependencies and get HEAD version of the main repo. The result is slightly inaccurate for the main repo, so we show a warning message that we do not know version of the main repo. In this case, we can just check HEAD version.
Despite the limitation, the overall result is good-enough, I have a very good idea of which licenses are used in the binary and only in rare cases will there be a mismatch. From what I heard, when a project changes license, most of the cases, it's switching from a permissive license to a more restrictive license, so analyzing HEAD version of the main repo is unlikely to give false negatives. False positives are easy to deal with, because people can just check them manually.

wlynch commented on v2/gocli/go_binary_test.go:
```diff
@@ -0,0 +1,94 @@
+			os.Chdir(filepath.Join(originalWorkDir, tc.workdir))
+			// This outputs the built binary as name "main".
+			binaryName := "main"
+			cmd := exec.Command("go", "build", "-o", binaryName)
```
I noticed when playing around with the tests that the test is not cleaning up after itself. Lets output binaries to a temp directory with [ioutil.TempDir](https://pkg.go.dev/io/ioutil#TempDir) and defer cleanup.

Bobgy replied:
done

wlynch commented on v2/gocli/go_binary.go:
```diff
@@ -0,0 +1,111 @@
+			err = fmt.Errorf("listModulesInGoBinary(path=%q): %w", path, err)
+		}
+	}()
+	depsBuildInfo, err := lichenmodule.Extract(ctx, path)
```
Let's add a TODO to replace this with the x/mod equivalent from https://github.com/golang/go/issues/39301 when it is available.

Bobgy replied:
added TODO

wlynch commented on v2/third_party/uw-labs/lichen/buildinfo/parse.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+// Parse parses build info details as returned by `go version -m [bin ...]`
+func Parse(info string) ([]model.BuildInfo, error) {
```
Lifted this idea from https://github.com/golang/go/issues/39301 - but Go has an equivalent impl in https://cs.opensource.google/go/go/+/refs/tags/go1.16.6:src/runtime/debug/mod.go;l=37. We'd have to make some modifications, but this would be a trimmed down library direct from the Go team.

Since this is likely going to be the impl that https://github.com/golang/go/issues/39301 is based off of, thoughts on using this over lichen?

Bobgy replied:
Fair enough, I'll use this version of code instead.

Bobgy replied:
Done

wlynch commented on v2/gocli/list.go:
```diff
@@ -0,0 +1,74 @@
+		// Note, we specifically want to replace version field.
+		// Haven't confirmed, but we may also need to override the
+		// entire struct when using replace directive with local folders.
+		if m.Replace != nil {
+			m = *m.Replace
+		}
```
My gut reaction is to preserve the Replace data, and handle replace overrides later on when needed. The doc comments don't really explain that this is happening, so this behavior might be confusing unless you're digging into the impl.

Bobgy replied:
I can see the confusion that, the data structure packages.Module we return contains the replace field, but we replace modules by default, so the field becomes useless.
Therefore, I added a Module type in this package, it no longer has the replace field, and I commented in related places about the behavior for replace directive. I insist in hiding the replace field, because it's completely an implementation details for the use-case of licensing, we should always check the replaced module instead.

What do you think?

wlynch replied:
SGTM!

wlynch commented:
> Therefore, I think I'd prefer staying with existing implementation for loading module metadata.

That's annoying. :(
Hopefully https://github.com/golang/go/issues/39301 will help out here.

Using CLI based impl in the meantime SGTM!

Bobgy commented:
Thank you for the review!
I will update soon

Bobgy commented:
Hi @wlynch, I've addressed all the feedback except the following:

I have some different opinions for two of them:
* https://github.com/google/go-licenses/pull/71#discussion_r689028527
* https://github.com/google/go-licenses/pull/71#discussion_r681714034

wlynch reviewed (commented):
Thanks for the changes - looking good!

wlynch reviewed (approved):
Simplified packages.Load looks great! 😄 

I'm still skeptical of the binary based approach - solely relying on source code seems like the better choice since we're dependent on the source's go.mod. We can't make guarantees about the binary matching the source, and documenting that this tool must only be ran with binaries that were built from the same source state as the current directory feels brittle and restrictive.

That said, I'm going to go ahead and merge this as-is. Since this is library-only and still in an experimental state (e.g. not hooked up into any CLI surface yet), we have flexibility to play around with this and try out ideas.

Thanks for your patience throughout the review. 🙏

wlynch commented:
Last thing is let's just squash these commits down for a clean commit message before merge.

Bobgy commented:
Awesome! Thank you for the flexibility, I am happy to contribute to this.

Let me squash the commits : )

Bobgy commented:
Hmm, it seems I cannot just reopen the PR after fixing my branch.
Blocked by error message "The branch was force-pushed or recreated".

I have to create a new PR for this.

Bobgy commented:
Recreated as https://github.com/google/go-licenses/pull/82
